internal/controller/biosversion_controller.go

```diff
@@ -0,0 +1,911 @@
+		}
+	}
+	// if already deleted or we deleted it or its created by user, remove reference
+	if apierrors.IsNotFound(err) || err == nil {
+		err = r.patchMaintenanceRequestRefOnBiosVersion(ctx, log, biosVersion, nil)
+		if err != nil {
+			return fmt.Errorf("failed to clean up serverMaintenance ref in BIOSVersionReconciler status: %w", err)
+		}
+	}
+	return nil
+}
+
+func (r *BIOSVersionReconciler) reconcile(
+	ctx context.Context,
+	log logr.Logger,
+	biosVersion *metalv1alpha1.BIOSVersion,
+) (ctrl.Result, error) {
+	if shouldIgnoreReconciliation(biosVersion) {
+		log.V(1).Info("Skipped BIOS Version reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	if modified, err := clientutils.PatchEnsureFinalizer(ctx, r.Client, biosVersion, biosVersionFinalizer); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+
+	return r.ensureBiosVersionStateTransition(ctx, log, biosVersion)
+
+}
+
+func (r *BIOSVersionReconciler) ensureBiosVersionStateTransition(
+	ctx context.Context,
+	log logr.Logger,
+	biosVersion *metalv1alpha1.BIOSVersion,
+) (ctrl.Result, error) {
+	server, err := r.getReferredServer(ctx, log, biosVersion.Spec.ServerRef)
+	if err != nil {
+		log.V(1).Info("referred server object could not be fetched")
+		return ctrl.Result{}, err
+	}
+
+	switch biosVersion.Status.State {
+	case "", metalv1alpha1.BIOSVersionStatePending:
+		return r.checkVersionAndTransistionState(ctx, log, biosVersion, server)
+	case metalv1alpha1.BIOSVersionStateInProgress:
+		if biosVersion.Spec.ServerMaintenanceRef == nil {
+			if requeue, err := r.requestMaintenanceOnServer(ctx, log, biosVersion, server); err != nil || requeue {
+				return ctrl.Result{}, err
+			}
+		}
+
+		if server.Status.State != metalv1alpha1.ServerStateMaintenance {
+			log.V(1).Info("Server is not in maintenance. waiting...", "server State", server.Status.State, "server", server.Name)
+			return ctrl.Result{}, nil
+		}
+
+		if server.Spec.ServerMaintenanceRef == nil || server.Spec.ServerMaintenanceRef.UID != biosVersion.Spec.ServerMaintenanceRef.UID {
+			// server in maintenance for other tasks. or
+			// server maintenance ref is wrong in either server or biosSettings
+			// wait for update on the server obj
+			log.V(1).Info("Server is already in maintenance for other tasks", "Server", server.Name, "serverMaintenanceRef", server.Spec.ServerMaintenanceRef)
+			return ctrl.Result{}, nil
+		}
+
+		return r.handleUpgradeInProgressState(ctx, log, biosVersion, server)
+	case metalv1alpha1.BIOSVersionStateCompleted:
+		// clean up maintenance crd and references and mark completed if version matches.
+		return r.checkVersionAndTransistionState(ctx, log, biosVersion, server)
+	case metalv1alpha1.BIOSVersionStateFailed:
+		log.V(1).Info("Failed to upgrade biosVersion", "ctx", ctx, "biosVersion", biosVersion, "server", server)
```

Comment:
Can we use `BIOSVersion` in all the logs - basically the same naming scheme as the resource type.